### ISSUE TITLE
About the validation code

### ISSUE THREAD

**Issue by Wangduo428** [Sompote/SAM3_LoRA#37]
Excellent project! I tried to validate the trained model on my own dataset but an error occurred on "create_coco_gt_from_dataset" function. It seems that the "idx" from the input "image_ids" is out of the range of dataset when doing "datapoint = dataset[idx]". I checked the code and was confused about line 976 to 985, as below:

                batch_size_actual = final_outputs['pred_logits'].shape[0]

                for i in range(batch_size_actual):
                    img_id = batch_idx * batch_size + i
                    all_image_ids.append(img_id)
                    all_predictions.append({
                        'pred_logits': final_outputs['pred_logits'][i].detach().cpu(),
                        'pred_boxes': final_outputs['pred_boxes'][i].detach().cpu(),
                        'pred_masks': final_outputs['pred_masks'][i].detach().cpu()
                    })

what does the "batch_size_actual" mean? Based on my understanding, "final_outputs['pred_logits'].shape[0]" refers to the number of prompts. Since one test image may have many prompts, why img_id increases with "batch_size_actual" as the code reads: "img_id = batch_idx * batch_size + i"?

Does it mean that this code only fits one image with only one prompt? For images with multiple prompts, how should I revise the code?

**Comment by Sompote** [comment]
Thank you for the careful analysis — you diagnosed it correctly, and it's now fixed in commit 54b2940. 🙏

You're right about this block:

batch_size_actual = final_outputs['pred_logits'].shape[0]

for i in range(batch_size_actual):
    img_id = batch_idx * batch_size + i
    all_image_ids.append(img_id)
    all_predictions.append({
        'pred_logits': final_outputs['pred_logits'][i].detach().cpu(),
        'pred_boxes': final_outputs['pred_boxes'][i].detach().cpu(),
        'pred_masks': final_outputs['pred_masks'][i].detach().cpu()
    })

What batch_size_actual really is. pred_logits.shape[0] is the number of text prompts (find-queries) in the batch, not the number of images. SAM3 batches per-prompt: the dataset builds one query per category, and the collator flattens all queries from all images into a single leading dimension. So batch_size_actual == total (image, prompt) pairs in the batch.

Why it broke. img_id = batch_idx * batch_size + i assumes exactly one prompt per image. That only holds for single-category data, where prompts-per-batch == images-per-batch == batch_size. With multi-category images, batch_size_actual > batch_size, so i runs past batch_size, img_id exceeds len(dataset), and you hit the out-of-range crash in create_coco_gt_from_dataset at dataset[idx]. Even within range, predictions get assigned to the wrong image. So yes — the old code was only correct for one-prompt-per-image (single-category) datasets.

The fix (pushed). The collator already preserves each prompt's source image and target objects, in the same flattened order as the model output. The loop now reads them from the batch instead of reconstructing positionally:

num_prompts = final_outputs['pred_logits'].shape[0]
meta = input_batch.find_metadatas[0]   # stage 0 (query_processing_order=0)
tgt  = input_batch.find_targets[0]

for i in range(num_prompts):
    eval_id = len(all_predictions)              # unique id per (image, prompt)
    src_idx = int(meta.original_image_id[i])    # real dataset index
    n       = int(tgt.num_boxes[i])
    obj_ids = tgt.object_ids_padded[i][:n].tolist() if n > 0 else []
    all_image_ids.append(eval_id)
    all_eval_units.append({'eval_id': eval_id, 'src_idx': src_idx, 'obj_ids': obj_ids})
    all_predictions.append({
        'pred_logits': final_outputs['pred_logits'][i].detach().cpu(),
        'pred_boxes':  final_outputs['pred_boxes'][i].detach().cpu(),
        'pred_masks':  final_outputs['pred_masks'][i].detach().cpu(),
    })

Each (image, prompt) pair becomes its own evaluation unit with a unique COCO id, and a new create_coco_gt_per_prompt() builds that unit's GT from only the objects the prompt targets — so a crack prediction is never scored against hole boxes, and empty/negative prompts get an annotation-free unit. That makes multi-class validation report correct mAP.

To use it: git pull and re-run your validation command — no other changes needed on your side. (I verified the data flow and ordering statically but couldn't run it end-to-end here, so if you can sanity-check on your multi-class set and confirm, that'd be great.)
